cpp/src/replace/clamp.cu

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/types.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/replace.hpp>
+#include <cudf/detail/iterator.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+struct dispacth_clamp {
+    template <typename T, typename Predicate>
+    void apply_transform (column_device_view  input,
+                          mutable_column_device_view output,
+                          Predicate pred,
+                          cudaStream_t stream)
+    {
+        if (input.nullable()){
+            auto input_begin = cudf::experimental::detail::make_null_replacement_iterator<T>(input);
+            thrust::transform(rmm::exec_policy(stream)->on(stream),
+                              input_begin,
+                              input_begin+input.size(),
+                              detail::make_validity_iterator(input),
+                              output.begin<T>(),
+                              pred);
+        } else {
+            thrust::transform(rmm::exec_policy(stream)->on(stream),
+                              input.begin<T>(),
+                              input.end<T>(),
+                              output.begin<T>(),
+                              pred);
+        }
+    }
+
+    template <typename T>
+    std::enable_if_t<cudf::is_fixed_width<T>(), std::unique_ptr<cudf::column>>
+    operator()(column_view const& input,
+              scalar const& lo,
+              scalar const& hi,
+              rmm::mr::device_memory_resource* mr,
+              cudaStream_t stream) {
+        using ScalarType = cudf::experimental::scalar_type_t<T>;
+        auto lo_scalar = static_cast<ScalarType const*>(&lo);
+        auto hi_scalar = static_cast<ScalarType const*>(&hi);
+        auto lo_value = lo_scalar->value(stream);
+        auto hi_value = hi_scalar->value(stream);
+        // mask will not change
+        std::unique_ptr<column> output {nullptr};
+        if (input.nullable()){
+            output = make_fixed_width_column(input.type(), input.size(),
+                                             copy_bitmask(input, stream, mr),
+                                             input.null_count(), stream, mr);
```

Comment:
Won't calling input.null_count() possibly invoke a kernel unnecessarily? Since you will be potentially changing  the null count, may want to leave the null count unknown.

Reply:
null mask or count will not change, that's why I am copying null mask. But, using `UNKNOWN_NULL_COUNT` would reduce burden on the algorithm.

Reply:
Good point. I'm not sure my suggestion was correct. @jrhemstad what should an algorithm do if it wants the output null count to be the same as the input null count? What if the input null count is unknown? Should we check that and copy that state to avoid unnecessary computation? Or should we copy the null count, potentially causing a kernel invocation (whose result may never be used)?